[2단계 - 로또 리팩토링] 듀이(이주희) 미션 제출합니다.

ljhee92 opened the pull request:
안녕하세요 아서! 듀이입니다.
주말 잘 보내고 계신가요?
주말에도 리뷰 요청을 드리게 되어 죄송하지만 편하신 시간에 천천히 봐주시면 감사드리겠습니다!😉

우선 2단계 요구사항이 MVC 패턴을 기반으로 리팩토링하는 것이었기 때문에 최대한 핵심 비즈니스 로직인 domain을 분리하려고 노력했습니다.
아래와 같은 내용들을 고민하면서 리팩토링하려고 했으니, 참고하여 리뷰 부탁드립니다!🫡

### 1. domain에 책임 부여
- ``Application``에서 검증하고 있던 당첨 번호 개수, 중복 검증 로직을 ``WinningNumbers`` 내부로 옮겼습니다.
- ``LottoManager``와 ``LottoMachine``의 역할이 겹쳐 depth가 깊어진 듯 하여 ``LottoManager``를 제거하고, 로또를 발행하는 책임을 ``LottoMachine``에게 부여했습니다.
- ``NumberGenerator`` 객체를 생성하여 로또 생성을 위한 랜덤 숫자 생성을 담당하게 하였습니다.
- 1단계에서 테스트하기 어려워 했던 로또 번호와 당첨 번호를 비교하는 검증 로직을 ``WinningNumbers`` 내부로 이동했습니다.
추가로 로또 번호와 보너스 번호를 비교하는 검증은 ``Lotto`` 내부로 이동시켰는데, 어쩄든 보너스 번호도 당첨과 연관되어있는 것이기 때문에 ``Lotto`` 내부에서 검증하는 것이 적절한 것인지 의문입니다. 이 내용에 대해 아서의 의견도 궁금합니다.

### 2. format 처리
- 1단계에서 페어와 진행할 때 페어는 ``formatted()`` 메서드를 사용하고 저는 ``String.format()`` 메서드를 사용하였는데, 출력하는 곳에서는 ``System.out.printf``로 통일해주었습니다.
- 저는 주로 정적 메서드인 ``String.format()``을 사용했었는데, 페어가 사용하던 ``formatted()``가 궁금해서 찾아보니 Java15에서 새로 추가된 인스턴스 메서드더라구요. 아서는 주로 포맷팅 시 어떻게 처리하시나요?

### 3. controller 추가
- 1단계에서는 굳이 controller를 도입하지 않아도 된다고 생각했는데, 2단계 준비를 하며 MVC 패턴을 왜 사용하는지 먼저 알아보았습니다.
- 주로 코드의 역할 분리, 유지 보수, 확장성, 재사용성 등을 위해 사용한다고 하는데, 이러한 이유들보다 ``Application`` 클래스는 프로그램의 실행만을 담당해야 한다는 이유에 설득당해 도입하게 됐습니다.
- 그렇게 controller를 만들었지만 여전히 controller의 역할이 무엇인가? 에 대한 의문에 대한 답을 명확하게 찾지 못했는데요. 아서의 의견도 궁금합니다!

### 4. 단위 테스트
- 최대한 분리했던 domain에 대해 테스트를 하려고 노력했습니다만, ``LottoMachine``의 ``calculateStatistics()`` 메서드는 테스트하기가 어렵더라구요.
- 당첨 통계를 가진 WinningStatistics를 생성하여 반환하는 아래 메서드인데, extracting으로 테스트해도 될지 어떻게 테스트해야할지 답을 찾지 못했습니다 조금만 도와주세요🥲
```
public static WinningStatistics calculateStatistics(final List<Lotto> lottos, final WinningNumbers winningNumbers,
                                                        final int bonusNumber) {
        Map<Rank, Integer> statistics = new EnumMap<>(Rank.class);
        for (final Lotto lotto : lottos) {
            int matchCount = winningNumbers.calculateMatchCount(lotto.getNumbers());
            boolean hasBonusNumber = lotto.hasNumber(bonusNumber);
            Rank rank = Rank.getRank(matchCount, hasBonusNumber);
            statistics.put(rank, statistics.getOrDefault(rank, 0) + 1);
        }
        return new WinningStatistics(statistics);
}
```

ljhee92 committed:
chore: domain 패키지 이동

ljhee92 committed:
refactor: 상수와 상태, 상수와 생성자 개행

ljhee92 committed:
refactor: 당첨 번호 검증 위치 변경
- 당첨 번호 검증을 WinningNumbers 객체 내부에서 하도록 위치를 변경하였다.

ljhee92 committed:
refactor: LottoManager 제거, PurchaseAmount 생성
- Lotto를 발행하는 것은 LottoMachine에게 역할을 위임하였다.
- 하나의 메서드는 하나의 일만 하기 위해 구입 금액을 입력 받는 메서드와 로또를 발행하는 메서드를 분리하였다.
- 구입 금액은 PurchaseAmount 객체를 생성하여 내부에서 검증하도록 하고, 로또 발행 장수를 반환할 수 있도록 하였다.

ljhee92 committed:
refactor: 숫자 생성을 LottoMachine이 아닌 NumbersGenerator에게 위임
- 로또의 생성은 LottoMachine에서 하되, 중복되지 않은 6개의 숫자를 생성하는 역할은 NumbersGenerator 객체를 생성하여 위임하였다.

ljhee92 committed:
refactor: Prize -> Rank 네이밍 변경, 여러 개의 if문 제거
- 로또 당첨 등수를 알아보기 쉽게 Prize -> Rank로 네이밍 변경하였다.
- 등수가 추가될 경우를 고려하여 여러 개의 if문을 제거하고 stream으로 비교하여 Rank를 반환할 수 있도록 변경하였다.

ljhee92 committed:
refactor: 로또 번호와 당첨 번호, 보너스 번호 비교 검증 로직 위치 변경
- 로또 번호와 당첨 번호 비교 검증 로직 위치를 LottoMachine이 아니라 WinningNumbers 내부로 변경하였다.
- 로또 번호와 보너스 번호 비교 검증 로직 위치를 Lotto 내부로 변경하였다.

ljhee92 committed:
refactor: 당첨 개수별 당첨 금액 계산 Rank 내부로 이동, 자동 반올림 버림 처리
- 당첨 개수별 당첨 금액을 계산하는 로직을 Rank 내부로 이동하였다.
- String.format의 %.2f가 자동 반올림 처리하기 때문에 WinningStatistics에서 수익률을 반환할 때 버림처리 해주었다.

ljhee92 committed:
refactor: formatted, String.format -> printf로 통일
- OutputView의 format 형식을 통일시켜주었다.

ljhee92 committed:
test: 당첨 통계 테스트 추가

ljhee92 committed:
test: Lotto, PurchaseAmount 테스트 추가
- 추가한 메서드에 대한 테스트를 추가하였다.

ljhee92 committed:
chore: controller 추가
- main은 프로그램의 실행만 담당할 수 있도록 controller를 추가하였다.

Hyunta commented on src/main/java/lotto/domain/PurchaseAmount.java:
```diff
@@ -0,0 +1,31 @@
+package lotto.domain;
+
+public class PurchaseAmount {
+    public static final int LOTTO_UNIT_PRICE = 1_000;
+    private static final int MAX_PURCHASE_AMOUNT = 100_000;
```
상수마다 public, private을 분리하시는 기준이 있으신가요?

ljhee92 replied:
다른 클래스에서도 사용하는 경우에 public으로 선언해 사용했습니다.
관리해야 하는 상수가 많아진다면 별도 클래스로 빼서 사용하는 것도 좋아 보이는데, 이번 미션에서는 굳이 필요성을 느끼지 못해서 다른 클래스에서 필요한 경우만 public으로 선언하였습니다.

Hyunta replied:
규칙에 일관성을 없는 것은 코드를 작성할 때와 변경점이 생겼을 때 단점으로 다가옵니다.

듀이와 제가 같이 개발을 하고 있다고 상황을 가정해볼게요.
`PurchaseAmount` 하위에 public 상수로 `LOTTO_UNIT_PRICE` 가 있는지 모르고, 제가 `Lotto` 클래스에 동일한 내용으로 상수를 만들었습니다.
기획자가 단위 금액을 1000원에서 1500원으로 변경해달라고 했을 때 실수로 Lotto 클래스에서만 수정하면 원하는데로 동작하지 않겠죠.

문제점
1. 다른 개발자가 상수를 새로 만들 때 관련된 클래스의 중복된 상수가 없는지 확인해야 합니다.
2. 변경 사항이 생겼을 때 누락하는 경우가 발생할 수 있습니다.

이러한 일이 벌어진데는 여러가지 원인이 있겠지만 상수 선언하는 형태에 일관성을 가지면 방지할 수 있다고 생각됩니다.

사실 미션에서의 코드 규모는 매우 작기 때문에 큰 의미를 못 찾을 수 있습니다. 그런데 결국 복잡한 환경에서 코드 유지보수를 잘하기 위한 훈련을 하고 있기 때문에 마치 거대한 프로젝트의 일부라고 생각하고 작성하시는 습관을 들이시면 도움되실 것 같습니다.

Hyunta commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,58 @@
+        OutputView.printWinningStatistics(winningStatistics, returnRate);
+    }
+
+    private static PurchaseAmount getPurchaseAmount() {
+        try {
+            int inputPurchaseAmount = InputView.inputPurchaseAmount();
+            return new PurchaseAmount(inputPurchaseAmount);
+        } catch (final IllegalArgumentException e) {
+            OutputView.printErrorMessage(e.getMessage());
+            return getPurchaseAmount();
+        }
+    }
```
해당 메서드을 static으로 선언해주신 이유가 있을까요?

ljhee92 replied:
앗,,, main에서 controller로 복사해오면서 static을 지우지 않았네요🥲
세심한 지적 감사합니다!

Hyunta commented on src/main/java/lotto/domain/PurchaseAmount.java:
```diff
@@ -0,0 +1,31 @@
+        this.purchaseAmount = purchaseAmount;
+    }
+
+    private void validatePurchaseAmount(final int purchaseAmount) {
+        if (purchaseAmount % LOTTO_UNIT_PRICE != 0) {
+            throw new IllegalArgumentException("구입금액은 %d원으로 나누어져야 합니다.".formatted(LOTTO_UNIT_PRICE));
+        }
+
+        if (purchaseAmount <= 0) {
+            throw new IllegalArgumentException("구입금액은 양수여야 합니다.");
+        }
+
+        if (purchaseAmount > MAX_PURCHASE_AMOUNT) {
+            throw  new IllegalArgumentException("구입금액은 최대 %d원까지입니다.".formatted(MAX_PURCHASE_AMOUNT));
+        }
+    }
```
예외 처리 메세지를 domain 패키지 하위에 위치시키신 이유가 있을까요?
내용이 view 에서 바로 보여지는 내용인 것 같네요

ljhee92 replied:
예외를 던지는 코드와 메시지를 관성적으로 함께 작성했었습니다🥲
아서 말씀대로 내용이 view에서 바로 보여지는 내용인 점은 생각을 못했었네요.
해결 방법으로 예외 메시지를 Enum으로 별도로 분리하는 방법이 적합할까요?

Hyunta replied:
enum으로 분리한다 하더라도 결국 해당 영역은 MVC 로 나눴을 때 Model에 있던 내용을 View에서 사용하고 있는 경우라고 생각됩니다. 그래서 어떤 의도를 가지고 Model의 예외 메세지를 view 단으로 전달하는지를 생각해보시면 좋을 것 같네요.
Model에 있는 값들 일부는 view 단으로 전달해도 괜찮지 않을까요?
view 와 model을 완전히 격리시키는 것이 어떤 의미를 갖는지 가볍게 고민해보시고 가능한지도 생각해보세요

ljhee92 replied:
model과 view를 완전히 격리시킨다는 것은 결국 model은 비즈니스 로직에 집중하게 하고 역할을 분리하기 위해서 라고 생각되는데요.
생각해보니 Enum으로 분리한다 하더라도 여전히 model에서 예외 메시지를 view에 넘기고 있긴 하네요🥲
분리한 Enum에 예외 상태코드와 메시지를 함께 관리하고 어떤 예외인지 커스텀 예외를 만들어 model에서는 예외의 상태만 포함하여 의도를 나타내고, view에서는 예외 상태에 따른 메시지를 표시하는 방법도 생각해볼 수 있을 것 같은데요!
아서가 저에게 생각해보라고 이야기해주신 부분이 이런 의도가 맞으셨을지 제가 잘 해석한게 맞는지 모르겠네요😅

Hyunta commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+import java.util.Map;
+
+public class LottoMachine {
+    private LottoMachine() {}
+
+    public static List<Lotto> issueLottos(final int lottoAmount) {
+        List<Lotto> lottos = new ArrayList<>();
+        for (int i = 0; i < lottoAmount; i++) {
+            Lotto lotto = new Lotto(NumbersGenerator.generateLottoNumbers());
+            lottos.add(lotto);
+        }
+        return lottos;
+    }
```
LottoMachine 클래스를 static 하게 사용하시게 된 이유가 궁금하네요ㅎㅎ

Hyunta replied:
아직 스트림이 익숙치 않으신 것 같은데 
해당 부분은 쉽게 한번 도전해볼 수 있을 것 같아서
stream 적용을 시도해보셔도 좋을 것 같아요.

ljhee92 replied:
> LottoMachine 클래스를 static 하게 사용하시게 된 이유가 궁금하네요ㅎㅎ

``LottoMachine``은 상태를 가지지 않는 객체라고 생각해서 static하게 사용했습니다.

> 아직 스트림이 익숙치 않으신 것 같은데
> 해당 부분은 쉽게 한번 도전해볼 수 있을 것 같아서
> stream 적용을 시도해보셔도 좋을 것 같아요.

넵 해당 부분 스트림 적용시켜보도록 하겠습니다!

Hyunta commented on src/main/java/lotto/domain/LottoMachine.java:
```diff
@@ -0,0 +1,31 @@
+        return lottos;
+    }
+
+    public static WinningStatistics calculateStatistics(final List<Lotto> lottos, final WinningNumbers winningNumbers,
+                                                        final int bonusNumber) {
+        Map<Rank, Integer> statistics = new EnumMap<>(Rank.class);
+        for (final Lotto lotto : lottos) {
+            int matchCount = winningNumbers.calculateMatchCount(lotto.getNumbers());
+            boolean hasBonusNumber = lotto.hasNumber(bonusNumber);
+            Rank rank = Rank.getRank(matchCount, hasBonusNumber);
+            statistics.put(rank, statistics.getOrDefault(rank, 0) + 1);
+        }
+        return new WinningStatistics(statistics);
+    }
```
`LottoMachine`이라는 이름을 들었을 때 저는 로또를 만드는 기계라고 생각했어요.
그런데 `WinninStatistics` 를 생성하는 책임까지 가지고 있네요.
해당 부분은 `WinningStatistics` 에서 처리하면 응집도가 높아지지 않을까요?

Hyunta replied:
> 최대한 분리했던 domain에 대해 테스트를 하려고 노력했습니다만, LottoMachine의 calculateStatistics() 메서드는 테스트하기가 어렵더라구요.
당첨 통계를 가진 WinningStatistics를 생성하여 반환하는 아래 메서드인데, extracting으로 테스트해도 될지 어떻게 테스트해야할지 답을 찾지 못했습니다 조금만 도와주세요🥲

WinningStatistics를 여기서 생성하는게 적절할 지 한번 고민해보시면 좋을 것 같습니다.
테스트하기 어려워진 이유가 무엇일까요?

ljhee92 replied:
어쩐지 저도 1단계 이후 리팩토링 진행하면서 ``LottoMachine``에서 당첨 통계를 계산하는게 맞을까? 하는 생각을 계속 했었어요! 아서 말씀대로 ``WinningStatistics``를 생성하는 책임은 ``WinningStatistics`` 객체 내부에서 처리한다. 라고 생각하니 테스트하기도 훨씬 쉬워졌네요! 반영해두었습니다😄

Hyunta commented on src/main/java/lotto/domain/WinningNumbers.java:
```diff
@@ -0,0 +1,39 @@
+import static lotto.domain.Lotto.LOTTO_SIZE;
+import static lotto.domain.Lotto.validateLottoNumber;
+
+public class WinningNumbers {
```
> 1단계에서 테스트하기 어려워 했던 로또 번호와 당첨 번호를 비교하는 검증 로직을 WinningNumbers 내부로 이동했습니다.
추가로 로또 번호와 보너스 번호를 비교하는 검증은 Lotto 내부로 이동시켰는데, 어쩄든 보너스 번호도 당첨과 연관되어있는 것이기 때문에 Lotto 내부에서 검증하는 것이 적절한 것인지 의문입니다. 이 내용에 대해 아서의 의견도 궁금합니다.

일치하는 로또 번호 개수는 `WinningNumbers`가 계산하고, 
보너스 번호는 Lotto에서 검증하는 것이 어색하다고 느껴지셔서 질문을 주신 것으로 이해됩니다. 
같이 어떤 일을 해야하는지를 한번 정리해보죠.

1. 구입한 로또와 당첨 로또 사이에 일치하는 갯수 계산
2. 보너스 번호가 일치하는지 계산

1번을 수행하기 위해서 생성하신 `WinnngNumbers` 객체를 사용했습니다.
2번을 수행하기 위해서 객체를 찾아보니 `BonusNumber` 가 아니라 int로 해당 값을 받았기 때문에 적절한 객체가 없습니다.
저는 일관성을 중요하게 생각하기 때문에 불필요하더라도 2번에 BonusNumber 객체를 만들어서 검증하는 방식으로 구현했을 것 같습니다.

ljhee92 replied:
동의합니다! ``BonusNumber`` 객체가 구매한 로또 번호에 포함되는지 검증하는 것이 더 명확하게 와닿는 책임이라고 생각되네요. 감사합니다!☺️

Hyunta commented on src/main/java/lotto/domain/NumbersGenerator.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Random;
+import java.util.Set;
+
+import static lotto.domain.Lotto.LOTTO_SIZE;
+import static lotto.domain.Lotto.MAX_LOTTO_NUMBER;
```
상수를 static import 하게 되면 어떤 장단점이 있을까요?

ljhee92 replied:
static import의 장점은 가독성 향상, 단점은 가독성 저하라고 생각합니다.
클래스를 언급하지 않고 선언할 수 있기 때문에 코드의 길이가 줄어 가독성이 좋아질 수 있습니다.
하지만 동시에 클래스를 언급하지 않기 때문에 어느 클래스에서 온 것인지 명확하지 않아 가독성 저하를 줄 수 있습니다.
전역적으로 사용되고 한 눈에 읽었을 때 네이밍만으로 의미를 유추할 수 있다면 static import를 사용하는 게 좋다고 생각합니다!

Hyunta replied:
장단점이 가독성 향상과 저하라는 부분이 모순적인 것 같으면서도 적절하네요.
말해주신 것처럼 상수가 static import 되어있다면 클래스의 출처를 알 수 없기 때문에 가독성이 저하됩니다.
만약 모든 상수가 private이라면 문제가 없겠지만 현재 코드에는 public한 상수도 존재하기 때문에 이럴 때는 클래스 이름까지 같이 남기는게 적절하지 않을까 하는 의견드립니다ㅎㅎ

Hyunta commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -1,18 +1,18 @@
 
     public static void printLottos(final List<Lotto> lottos) {
-        System.out.println("%d개를 구매했습니다.".formatted(lottos.size()));
+        System.out.printf("%d개를 구매했습니다.%n", lottos.size());
```
> 저는 주로 정적 메서드인 String.format()을 사용했었는데, 페어가 사용하던 formatted()가 궁금해서 찾아보니 Java15에서 새로 추가된 인스턴스 메서드더라구요. 아서는 주로 포맷팅 시 어떻게 처리하시나요?

저도 보통 String.format() 을 썼었습니다.
실무에서는 거의 Rails를 써서 Java 포맷팅은 깊게 고민해본 적은 없었습니다ㅠ
그런데 formatted를 쓰면 변수 부분이 확실하게 구분되어져서 괜찮아보이네요ㅎㅎ

Hyunta commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,58 @@
+
+import static lotto.domain.Lotto.validateLottoNumber;
+
+public class LottoController {
```
> 그렇게 controller를 만들었지만 여전히 controller의 역할이 무엇인가? 에 대한 의문에 대한 답을 명확하게 찾지 못했는데요. 아서의 의견도 궁금합니다!

controller는 정확히 MVC 역할에 맞는 기능을 하도록 만들어야 된다고 생각할 필요는 없습니다.
MVC 패턴이라는 개념 자체가 조금 모호하기 때문에 이에 대한 뚜렷한 답변을 얻기는 어렵습니다.
지금 하신 것처럼 적절히 비즈니스 로직과 화면 구성단을 분리하고 그 사이에서 상호작용을 하는 역할 정도로 받아들이면 적절할 것 같습니다.
아래 글을 읽어봐도 좋을 것 같아요ㅎㅎ
https://velog.io/@eddy_song/mvc

Hyunta reviewed (changes requested):
안녕하세요 듀이,
2단계 미션도 잘 진행해주셨네요 👍🏻
몇가지 커멘트 남겨드렸으니 확인 부탁드립니다.

ljhee92 committed:
chore: 복사 시 제거하지 못한 static 제거

ljhee92 committed:
refactor: 로또 발행 for문 -> 스트림 변경

ljhee92 committed:
refactor: 전역 사용 상수 Enum 처리
- view에서 바로 보여지는 예외 처리 메시지가 domain 내부에 그대로 위치하고 있었으므로 Enum으로 분리하였다.
- 전역적으로 사용되는 상수를 Enum으로 분리하였다.

ljhee92 committed:
refactor: BonusNumber 객체 생성
- 보너스 번호가 구매한 로또 번호에 포함되는지 검증하는 책임을 BonusNumber 객체를 생성하여 위임하였다.

ljhee92 committed:
refactor: 당첨 통계 계산 로직을 WinningStatistics 내부로 위임
- LottoMachine은 로또를 발행하는 책임만 갖게 하고, 당첨 통계를 계산하는 로직은 WinningStatistics 객체에게 위임하였다.

Hyunta reviewed (approved):
안녕하세요 듀이,
2단계 리팩터링 미션도 잘 진행해주셨네요.
드렸던 질문들이 앞으로 미션을 진행하시면서 도움이 되었으면 합니다.
몇가지 추가 커멘트 남겨드렸는데 궁금하신 부분이 있으면 편하게 DM 주세요.
